server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.node;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.search.TransportSearchAction;
+import org.elasticsearch.action.support.PlainActionFuture;
+import org.elasticsearch.action.support.RefCountingListener;
+import org.elasticsearch.action.support.SubscribableListener;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.http.HttpServerTransport;
+import org.elasticsearch.index.reindex.ReindexAction;
+import org.elasticsearch.node.internal.TerminationHandler;
+import org.elasticsearch.tasks.TaskManager;
+
+import java.util.ArrayList;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.core.Strings.format;
+
+/**
+ * This class was created to extract out the logic from {@link Node#prepareForClose()} to facilitate testing.
+ * <p>
+ * Invokes hooks to prepare this node to be closed. This should be called when Elasticsearch receives a request to shut down
+ * gracefully from the underlying operating system, before system resources are closed.
+ * <p>
+ * Note that this class is part of infrastructure to react to signals from the operating system - most graceful shutdown
+ * logic should use Node Shutdown, see {@link org.elasticsearch.cluster.metadata.NodesShutdownMetadata}.
+ */
+public class ShutdownPrepareService {
+
+    private final Logger logger = LogManager.getLogger(ShutdownPrepareService.class);
+    Settings settings;
+    HttpServerTransport httpServerTransport;
+    TerminationHandler terminationHandler;
```

Comment:
We should make all fields as final and private as possible. I think these can be private final?

Reply:
Yes. Thanks!